src/include/duckdb/common/chimp/chimp_compress.hpp

```diff
@@ -0,0 +1,305 @@
+
+	ChimpState<T, false> state;
+
+public:
+	idx_t RequiredSpace() const {
+		idx_t required_space = ChimpPrimitives::MAX_BYTES_PER_VALUE;
+		// Any value could be the last,
+		// so the cost of flushing metadata should be factored into the cost
+		// byte offset of data
+		required_space += sizeof(byte_index_t);
+		// amount of leading zero blocks
+		required_space += sizeof(uint8_t);
+		// first leading zero block
+		required_space += 3;
+		// amount of flag bytes
+		required_space += sizeof(uint8_t);
+		// first flag byte
+		required_space += 1;
+		return required_space;
+	}
+
+	// How many bytes the data occupies for the current segment
+	idx_t UsedSpace() const {
+		return state.chimp_state.output.BytesWritten();
+	}
+
+	idx_t RemainingSpace() const {
+		return metadata_ptr - (handle.Ptr() + UsedSpace());
+	}
+
+	idx_t CurrentGroupMetadataSize() const {
+		idx_t metadata_size = 0;
+
+		metadata_size += 3 * state.chimp_state.leading_zero_buffer.BlockCount();
+		metadata_size += state.chimp_state.flag_buffer.BytesUsed();
+		metadata_size += 2 * state.chimp_state.packed_data_buffer.index;
+		return metadata_size;
+	}
+
+	// The current segment has enough space to fit this new value
+	bool HasEnoughSpace() {
+		auto space_used = AlignValue(UsedSpace() + RequiredSpace()) + metadata_byte_size + CurrentGroupMetadataSize();
+		// if (space_used + 50 >= Storage::BLOCK_SIZE / 2) {
+		//	return false;
+		// }
+		if (handle.Ptr() + ChimpPrimitives::HEADER_SIZE + AlignValue(UsedSpace() + RequiredSpace()) >=
+		    (metadata_ptr - CurrentGroupMetadataSize())) {
+			// printf("Header_size(4) + UsedSpace(%llu) + RequiredSpace(%llu) = %llu | MetadataSize = %llu |
+			// CurrentGroupMetadataSize = %llu\n", 	UsedSpace(), RequiredSpace(), ChimpPrimitives::HEADER_SIZE +
+			// AlignValue(UsedSpace() + RequiredSpace()), 	RemainingSpace(), CurrentGroupMetadataSize());
+			return false;
+		}
+		return true;
+	}
+
+	void CreateEmptySegment(idx_t row_start) {
+		group_idx = 0;
+		metadata_byte_size = 0;
+		auto &db = checkpointer.GetDatabase();
+		auto &type = checkpointer.GetType();
+		auto compressed_segment = ColumnSegment::CreateTransientSegment(db, type, row_start);
+		compressed_segment->function = function;
+		current_segment = move(compressed_segment);
+
+		auto &buffer_manager = BufferManager::GetBufferManager(db);
+		handle = buffer_manager.Pin(current_segment->block);
+
+		segment_data = handle.Ptr() + current_segment->GetBlockOffset() + ChimpPrimitives::HEADER_SIZE;
+		metadata_ptr = handle.Ptr() + current_segment->GetBlockOffset() + Storage::BLOCK_SIZE;
+		state.AssignDataBuffer(segment_data);
+		state.chimp_state.Reset();
+	}
+
+	void Append(UnifiedVectorFormat &vdata, idx_t count) {
+		auto data = (T *)vdata.data;
+
+		for (idx_t i = 0; i < count; i++) {
+			auto idx = vdata.sel->get_index(i);
+			state.template Update<ChimpWriter>(data[idx], vdata.validity.RowIsValid(idx));
+		}
+	}
+
+	void WriteValue(CHIMP_TYPE value, bool is_valid) {
```

Comment:
Might be worthwhile to make a loop that calls WriteValue x amount of times until the group is full, to make this a bit more efficient